### PR TITLE
Add Prometheus Metrics

### PR DESCRIPTION
**Pull request by arlyon**
This PR publishes information about the pusher and fetcher to the prometheus metrics endpoint.

What is left out of this is publishing qualifiers, which may be an interesting metric, but will be included in a subsequent PR

### CODE REVIEW

File: pkg/storage/asset_store.go [hunk 1]
```diff
@@ -11,21 +11,27 @@ import (
...
 // AssetStore is a wrapper around a BlobAccess to inteface well with
 // AssetReference messages
-type AssetStore struct {
+type AssetStore interface {
+	Get(ctx context.Context, ref *asset.AssetReference, instance digest.InstanceName) (*asset.Asset, error)
+	Put(ctx context.Context, ref *asset.AssetReference, data *asset.Asset, instance digest.InstanceName) error
+}
+
+// blobAccessAssetStore is an AssetStore backed by a blobAccess.
+type blobAccessAssetStore struct {
```
Comment (tomcoldrick-ct) [inline pkg/storage/asset_store.go:20]: This should be in its own file if we're making `AssetStore` an interface, imo, similarly the `NewAssetStore` should become `NewBlobAccessAssetStore`.

**Review by tomcoldrick-ct — COMMENTED** [review]
Nice one, thanks a lot! Overall I think this patch looks great, just a few nits (plus responses to your comments and some todos).

>    I have written an impl for metrics_asset_store but haven't hooked it up yet. Will we need that? From what I can tell publishing asset store metrics is somewhat covered by the others.

In my humble opinion this is adequately covered from the blobaccess metrics along with the `Fetch` and `Push` server metrics.

 >    I have published duration_seconds and size_bytes. I believe most other interesting things can be inferred. Is that thinking correct?

I _think_ qualifiers may be an additional vector we might want, given their flexibility. From an operator perspective I'd probably like to know if, e.g. my git qualified fetches are failing more often or whatever. That said, such a thing is probably obvious anyway. Another potential thing that crossed my mind while looking at this is some kind of metric to allow me to calculate "% cache hit", however in the current design this is kinda awkward. I suspect that redesign related to #3 will be informative on this, so it's probably best to wait for that in any case.

File: pkg/storage/asset_store.go (hunk 1 above)
Comment (arlyon) [reply pkg/storage/asset_store.go:20]: If metrics asset store is superfluous I will instead revert those changes.

Comment (arlyon) [reply pkg/storage/asset_store.go:20]: If metrics asset store is superfluous I will instead revert those changes.

File: pkg/fetch/metrics_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+func (mf *metricsFetcher) FetchDirectory(ctx context.Context, req *remoteasset.FetchDirectoryRequest) (*remoteasset.FetchDirectoryResponse, error) {
+	timeStart := mf.clock.Now()
+	resp, err := mf.fetcher.FetchDirectory(ctx, req)
+	if err != nil {
+		mf.updateDurationSeconds(mf.fetchDirectoryDurationSeconds, status.Code(err), timeStart)
+		return nil, err
+	}
+	mf.updateDurationSeconds(mf.fetchDirectoryDurationSeconds, codes.Code(resp.Status.Code), timeStart)
+	// todo(arlyon): maybe list qualifiers?
```
[the remark this answers is not in the document]
Comment (tomcoldrick-ct) [reply pkg/fetch/metrics_fetcher.go]: Opened #14 to track this, I don't think we should block on it here.

**Review by tomcoldrick-ct — APPROVED** [review]
Aside from the failing CI (looks like buildifier is not appeased), this looks good to me!

**Pull request merged**